### PR TITLE
Optimisation

### PR DESCRIPTION
**Pull request by fabientownsend**

### CODE REVIEW

File: lib/computer.ex [hunk 1]
```diff
@@ -3,53 +3,97 @@ defmodule Computer do
...
   def move(board, mark) do
+    if board == Board.empty_board do
+      random_move()
+    else
+      board |> most_valuable_move(mark)
+    end
+  end
+
+  def random_move() do
+    Enum.take_random([0, 2, 4, 6, 8], 1) |> Enum.at(0)
```
Comment (felipesere) [inline lib/computer.ex]: This can be simplified with https://hexdocs.pm/elixir/Enum.html#random/1

File: lib/computer.ex [hunk 2]
```diff
@@ -3,53 +3,97 @@ defmodule Computer do
...
+      positions_available,
+      fn(position) ->
+        {position, minimax(Board.set_mark(board, mark, position), mark, false, 9)}
+      end
+    )
+  end
+
+  def select_best_move(scored_moves) do
+    scored_moves
+    |> Enum.max_by(fn(move) -> elem(move, 1) end)
```
Comment (felipesere) [inline lib/computer.ex]: this cold be `Enum.max_by(fn {_, move}) ->move)`

File: lib/computer.ex [hunk 3]
```diff
@@ -3,53 +3,97 @@ defmodule Computer do
...
+    most_valuable_move.value_move
   end
 
-  defp minimizing_player(board, mark, depth) do
-    spots = board |> Board.available_placeholders
-    all_values = spots |> Enum.map(
-      fn(x) ->
-        new_board = Board.set_mark(board, other_mark(mark), x)
-        minimax(new_board, mark, true, depth - 1)
+  defp minimizing_player(board, mark, depth, alpha, beta) do
```
Comment (felipesere) [inline lib/computer.ex:78]: This duplication is pretty narly. Maybe you can extract the differences into a function that you pass in?

File: lib/computer.ex [hunk 4]
```diff
@@ -3,53 +3,97 @@ defmodule Computer do
...
+    |> try_all_moves(board, mark)
+    |> select_best_move
   end
 
-  defp minimax(board, mark) do
-    minimax(board, mark, false, 9)
+  def try_all_moves(positions_available, board, mark) do
+    Enum.map(
+      positions_available,
+      fn(position) ->
```
Comment (felipesere) [inline lib/computer.ex:27]: I don't think this reads particularly well...
position
|> Enum.map(...)

And then see if you can shrink that function? `false` and `9` seem to be constants!

**Review by felipesere — CHANGES_REQUESTED** [review]

File: lib/computer.ex (hunk 4 above)
Comment (fabientownsend) [reply lib/computer.ex:27]: About the 9,  the computer is really tied to a 3*3 board at the moment. 

Then random move isn't really random as it's the best opening move for a 3*3, I don't really know how to make it scale other by really randomise if the board is over 3*3.

File: lib/computer.ex (hunk 2 above)
Comment (fabientownsend) [reply lib/computer.ex]: Look nice

File: lib/computer.ex (hunk 3 above)
Comment (fabientownsend) [reply lib/computer.ex:78]: I thought to use a hash for the variable, _in a Ruby way_. I'm not really sure about the result and if it's really the Elixir way. I could give a try

File: lib/computer.ex (hunk 1 above)
Comment (fabientownsend) [reply lib/computer.ex]: 👍

File: lib/computer.ex (hunk 3 above)
Comment (felipesere) [reply lib/computer.ex:78]: I put the comment here, but I don't mean the actual signature... the duplication between `minimizing_player` and `maximizing_player` is what I'd prefer to see reduced

Comment (fabientownsend) [reply lib/computer.ex:78]: Yeah I was on the same page 👍

**Pull request merged**